File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isSpace, isLetter, isUpper)
import qualified Data.Text as T

responseFor :: T.Text -> T.Text
responseFor xs
  | isSilence = T.pack "Fine. Be that way!"
  | isYelling &&
    isInquiry = T.pack "Calm down, I know what I'm doing!"
  | isYelling = T.pack "Whoa, chill out!"
  | isInquiry = T.pack "Sure."
  | otherwise = T.pack "Whatever."
  where
    compact = T.filter (not.isSpace) xs
    letters = T.filter isLetter compact
    isSilence = T.null compact
    isYelling = not (T.null letters) && T.all isUpper letters
    isInquiry = T.last compact == '?'

File: ruby/bob/bob.rb
module Bob
  YELL = /(?=.*\p{Lu})\P{Ll}+/

  def self.hey(remark)
    case remark
    when /\A\s*\Z/ then "Fine. Be that way!"
    when /\A#{YELL}\?\s*\Z/ then "Calm down, I know what I'm doing!"
    when /\?\s*\Z/ then "Sure."
    when /\A#{YELL}\Z/ then "Whoa, chill out!"
    else "Whatever."
    end
  end
end
